File: app/src/main/java/gr/tsagi/jekyllforandroid/app/utils/ParsePostData.java
package gr.tsagi.jekyllforandroid.app.utils;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.util.Base64;
import gr.tsagi.jekyllforandroid.app.data.PostsContract.PostEntry;

import java.io.UnsupportedEncodingException;


/**
 * Created by tsagi on 8/15/14.
 */
public class ParsePostData {

    private final String LOG_TAG = ParsePostData.class.getSimpleName();
    private final Context mContext;

    public ParsePostData(Context context) {
        mContext = context;
    }


    final String JK_TITLE = "title";
    final String JK_CATEGORY = "category";
    final String JK_TAGS = "tags";


    public ContentValues getDataFromContent(String id, String contentBytes, int type) {

        // Get and insert the new posts information into the database
        String postContent = null;

        // Blobs return with Base64 encoding so we have to UTF-8 them.
        byte[] bytes = Base64.decode(contentBytes, Base64.DEFAULT);
        try {
            postContent = new String(bytes, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }

//        StringBuilder stringBuilder = new StringBuilder();
//
//        InputStream is;
//        BufferedReader r;
//
//        is = new ByteArrayInputStream(postContent.getBytes());
//        // read it with BufferedReader
//        r = new BufferedReader(new InputStreamReader(is));
//        String line;
//
//        int yaml_dash = 0;
//        String yamlStr = "";
//        try {
//            while ((line = r.readLine()) != null) {
//                if (line.equals("---")) {
//                    yaml_dash++;
//                }
//                if (yaml_dash < 2) {
//                    if (!line.equals("---"))
//                        yamlStr = yamlStr + line + "\n";
//                }
//                if (yaml_dash >= 2) {
//                    if (!line.equals("---"))
//                        if (line.equals(""))
//                            stringBuilder.append("\n");
//                        else
//                            stringBuilder.append(line);
//                }
//            }
//            is.close();
//        } catch (IOException e) {
//            e.printStackTrace();
//        }


//        String content = stringBuilder.toString();



//        Yaml yaml = new Yaml();

//        Map map;
//        map = (Map) yaml.load(yamlStr);

        String title=id;
        String tags="null";
        String category="null";

//        title = map.get(JK_TITLE).toString();

//        if (map.containsKey(JK_TAGS)) {
//            tags = map.get(JK_TAGS).toString().replace("[", "").replace("]", "");
//        } else {
//            tags = "null";
//        }
//        if (map.containsKey(JK_CATEGORY)) {
//            category = map.get(JK_CATEGORY).toString();
//        } else {
//            category = "null";
//        }

//        addTags(tags, id);
//        addCategory(category, id);

        long date = 0;

        if (type == 0) {
            int i = id.indexOf('-', 1 + id.indexOf('-', 1 + id.indexOf('-')));
            date = Long.parseLong(id.substring(0, i).replace("-", ""));
        }

        // First, check if the post exists in the db
        Cursor cursorId = mContext.getContentResolver().query(
                PostEntry.CONTENT_URI,
                new String[]{PostEntry.COLUMN_POST_ID},
                PostEntry.COLUMN_POST_ID + " = ?",
                new String[]{id},
                null);

//        Cursor cursorContent = mContext.getContentResolver().query(
//                PostEntry.CONTENT_URI,
//                new String[]{PostEntry.COLUMN_CONTENT},
//                PostEntry.COLUMN_CONTENT + " = ?",
//                new String[]{content},
//                null);

        ContentValues postValues = new ContentValues();

        if (cursorId.moveToFirst()) {
            cursorId.close();
//            if (!cursorContent.moveToFirst()) {
//                cursorContent.close();
                ContentValues updateValues = new ContentValues();
                updateValues.put(PostEntry.COLUMN_CONTENT, postContent);
                if (updateValues.size() > 0) {
                    mContext.getContentResolver().update(PostEntry.CONTENT_URI, updateValues,
                            PostEntry.COLUMN_POST_ID + " = \"" + id + "\"", null);
                }

//            }
        } else {
            cursorId.close();
//            cursorContent.close();

            postValues.put(PostEntry.COLUMN_TITLE, title);
            postValues.put(PostEntry.COLUMN_DATETEXT, date);
            postValues.put(PostEntry.COLUMN_DRAFT, type);
            postValues.put(PostEntry.COLUMN_CONTENT, postContent);
            postValues.put(PostEntry.COLUMN_POST_ID, id);
        }

        return postValues;

    }

}

File: app/src/main/java/gr/tsagi/jekyllforandroid/app/utils/ParsePostData.kt
package gr.tsagi.jekyllforandroid.app.utils

import android.util.Base64
import android.util.Log
import com.jchanghong.data.DatabaseManager
import com.jchanghong.model.Category
import com.jchanghong.model.Note
import com.jchanghong.utils.date_id_toTitle
import com.jchanghong.utils.getyam
import com.jchanghong.utils.path2Catogery
import org.yaml.snakeyaml.Yaml
import java.io.UnsupportedEncodingException


/**
\* Created with IntelliJ IDEA.
\* User: jchanghong
\* Date: 8/15/14
\* Time: 15:14
\*/
class ParsePostData {

    internal val JK_TITLE = "title"
    internal val JK_CATEGORY = "category"
    internal val JK_TAGS = "tags"

    fun getNoteFrombyte(id: String, path: String, contentBytes: String, type: Int): Note {

        // Get and insert the new posts information into the database
        val note = Note()
        // Blobs return with Base64 encoding so we have to UTF-8 them.
        val bytes = Base64.decode(contentBytes, Base64.DEFAULT)
        val postContent: String =
                try {
                    String(bytes, charset("UTF-8"))
                } catch (e: UnsupportedEncodingException) {
                    e.printStackTrace()
                    ""
                }
        note.content = postContent

//        val stringBuilder = StringBuilder()
//
//        val inputStream: InputStream
//        val r: BufferedReader
//
//        inputStream = ByteArrayInputStream(postContent?.toByteArray())
//        // read it with BufferedReader
//        r = BufferedReader(InputStreamReader(inputStream))
//        var line: String?
//
//        var yaml_dash = 0
//        var yamlStr = ""
//        try {
//            while (true) {
//                line = r.readLine()
//                if (line == null) {
//                    break
//                }
//                if (line == "---") {
//                    yaml_dash++
//                }
//                if (yaml_dash < 2) {
//                    if (line != "---")
//                        yamlStr = yamlStr + line + "\n"
//                }
//                if (yaml_dash >= 2) {
//                    if (line != "---")
//                        if (line == "")
//                            stringBuilder.append("\n")
//                        else
//                            stringBuilder.append(line)
//                }
//            }
//            inputStream.close()
//        } catch (e: IOException) {
//            e.printStackTrace()
//        }
//
//
//        val content = stringBuilder.toString()
//

        val yaml = Yaml()
        val map: Map<*, *>?
        map = yaml.load(postContent.getyam()) as? Map<*, *>
        var title:String
        var tags: String? = null
        val category: String?

        if (map == null) {
            note.tittle = id.date_id_toTitle()
            var date: Long = 0
            if (type == 0) {
                val i = id.indexOf('-', 1 + id.indexOf('-', 1 + id.indexOf('-')))
                date = java.lang.Long.parseLong(id.substring(0, i).replace("-", ""))
            }
            note.lastEdit = date
            note.category = getornew(null, path.path2Catogery())
            return note
        }
        title = map[JK_TITLE].toString()

        note.tittle = title
        if (map.containsKey(JK_TAGS)) {
            tags = map[JK_TAGS].toString().replace("[", "").replace("]", "")
        } else {
            tags = null
        }
        if (map.containsKey(JK_CATEGORY)) {
            category = map[JK_CATEGORY].toString()
        } else {
            category = null
        }
        val categorymode = getornew(tags, category)
        note.category = categorymode

        var date: Long = 0

        if (type == 0) {
            val i = id.indexOf('-', 1 + id.indexOf('-', 1 + id.indexOf('-')))
            date = java.lang.Long.parseLong(id.substring(0, i).replace("-", ""))
        }
        note.lastEdit = date
        // First, check if the post exists in the db
        return note
    }

    private fun getornew(tags: String?, category: String?): Category {
        if (category != null) {
            var c = DatabaseManager.getCategoryByName(category)

            if (c != null) {
                return c
            } else {
                c = Category(name = category)
                DatabaseManager.insertCategory(c)
                return c
            }
        }
        if (tags != null) {
            val c1 = Category(name = tags.trim().split(" ")[0])
            val c = DatabaseManager.getCategoryByName(c1.name)

            if (c != null) {
                return c
            } else {
                DatabaseManager.insertCategory(c1)
                return c1
            }
        }
        Log.d("jiangchanghong", "no cat find ,use default")
        return DatabaseManager.defaultCAT
    }

}
